src/util/git.js

```diff
@@ -385,89 +341,90 @@ export default class Git {
    */
   async init(): Promise<string> {
     this.gitUrl = await Git.secureGitUrl(this.gitUrl, this.hash, this.reporter);
+
+    await this.setRefRemote();
+
     // check capabilities
-    if (await Git.hasArchiveCapability(this.gitUrl)) {
+    if (this.ref !== '' && (await Git.hasArchiveCapability(this.gitUrl))) {
       this.supportsArchive = true;
     } else {
       await this.fetch();
     }
 
-    return this.setRefRemote();
+    return this.hash;
   }
 
   async setRefRemote(): Promise<string> {
-    const stdout = await Git.spawn(['ls-remote', '--tags', '--heads', this.gitUrl.repository]);
-    const refs = Git.parseRefs(stdout);
+    const stdout = await spawnGit(['ls-remote', '--tags', '--heads', this.gitUrl.repository]);
+    const refs = parseRefs(stdout);
+    return this.setRef(refs);
+  }
+
+  setRefHosted(hostedRefsList: string): Promise<string> {
+    const refs = parseRefs(hostedRefsList);
     return this.setRef(refs);
   }
 
   /**
-   * TODO description
+   * Resolves the default branch of a remote repository (not always "master")
    */
 
-  async setRef(refs: GitRefs): Promise<string> {
-    // get commit ref
-    const {hash} = this;
-
-    const names = Object.keys(refs);
-
-    if (Git.isCommitHash(hash)) {
-      for (const name in refs) {
-        if (refs[name] === hash) {
-          this.ref = name;
-          return hash;
-        }
-      }
-
-      // `git archive` only accepts a treeish and we have no ref to this commit
-      this.supportsArchive = false;
-
-      if (!this.fetched) {
-        // in fact, `git archive` can't be used, and we haven't fetched the project yet. Do it now.
-        await this.fetch();
-      }
-      return (this.ref = this.hash = hash);
-    }
-
-    const ref = await this.findResolution(hash, names);
-    const commit = refs[ref];
-    if (commit) {
-      this.ref = ref;
-      return (this.hash = commit);
-    } else {
-      throw new MessageError(this.reporter.lang('couldntFindMatch', ref, names.join(','), this.gitUrl.repository));
+  async resolveDefaultBranch(): Promise<ResolvedSha> {
+    try {
+      const stdout = await spawnGit(['ls-remote', '--symref', this.gitUrl.repository, 'HEAD']);
+      const lines = stdout.split('\n');
+      const [, ref] = lines[0].split(/\s+/);
+      const [sha] = lines[1].split(/\s+/);
```

Comment:
I'd rather use a stronger regexp like `/^\s*([0-9a-f]{5,40})\s+/` and just use `match` on it. Same goes for the `ref` if possible.